capsules/src/max17205.rs

```diff
@@ -261,21 +311,56 @@ impl<'a> i2c::I2CClient for MAX17205<'a> {
             State::ReadCurrent => {
                 let current = ((buffer[1] as u16) << 8) | (buffer[0] as u16);
 
+                let mut error = ReturnCode::SUCCESS;
+                if _error != i2c::Error::CommandComplete {
+                    error = ReturnCode::ENOACK;
+                }
+
                 self.client
                     .get()
-                    .map(|client| client.voltage_current(self.voltage.get(), current));
+                    .map(|client| client.voltage_current(self.voltage.get(), current, error));
 
                 self.buffer.replace(buffer);
                 self.i2c0.disable();
                 self.state.set(State::Idle);
             }
+            State::SetupReadRomID => {
+                self.i2c1.read(buffer, 8);
+                self.state.set(State::ReadRomID);
+            }
+            State::ReadRomID => {
+
+                let exists = self.buf.map_or(false, |app| {
+                    app.rom_id_buffer.as_mut().map(|dest_buf| {
+                        let d = &mut dest_buf.as_mut()[0..8];
+                        for (i, c) in buffer[0..8].iter().enumerate() {
+                            d[i] = *c;
+                        }
+                    });
+                    app.rom_id_buffer.is_some()
+                });
+
+                self.buffer.replace(buffer);
+
+                let mut error = ReturnCode::SUCCESS;
+                if _error != i2c::Error::CommandComplete {
+                    error = ReturnCode::ENOACK;
+                }
+
+                if !exists {
+                    error = ReturnCode::ENOMEM;
+                }
+
+                self.client.get().map(|client| client.romid(error));
```

Comment:
I would structure this differently, since this capsule is sort of doing double duty as an internal driver for the MAX17205 as well as exposing a user-facing system call interface.

I would push the logic of filling the allow buffer to the `Max17205Client#romid` method, and change it's signature to something like:

```rust
fn romid(&self, rid: u64, error: ReturnCode);
```

Then this arm of the match statement just looks something like:

```rust
// u64 rid from 8 bytes
let rid = buffer.iter().take(8).enumerate().fold(0u64, |rid, (i, b)| rid | (b as u64 << i * 8));
self.buffer.replace(buffer);

let error =  if _error == i2c::Error::CommandComplete {
    ReturnCode::SUCCESS;
else {
     ReturnCode::ENOACK;
}

self.client.get().map(|client| client.romid(rid, error));
```